README.md
[![License](https://img.shields.io/badge/License-MPL2-blue.svg)](https://github.com/hawkeye116477/S4E_Revived/blob/master/LICENSE)

-----------------

# [Status-4-Evar Revived](https://addons.basilisk-browser.org/addon/s4e-revived/) for Waterfox Classic and Basilisk

### Note: This extension is no longer under active development. Feel free to fork it under new extension ID and namespace!


ChangeLog
Version 2019.10.11.3
* Fixed install.rdf
* Fixed icon for advanced options

Version 2019.10.11.2
* Fixed install.rdf

Version 2019.10.11.1
* Restored compatibility with Waterfox Classic 2019.10

Version next
* Drop support for Firefox 51.0 and older.

Version 2017.08.20.15
* Consolidate locale handling.
* Replace deprecated syntax.

Version 2017.03.19.12b
* Fix Firefox 54+ compatibility (Bug 1335877) by Jonas Jenwald.

Version 2016.10.11.01

Version 2016.10.09.22b
* Fix throbber icon (Bug 750054).

Version 2016.10.09.21b
* Fix Firefox 49+ compatibility.

Version 2016.07.16.21b
* Drop support for Firefox 44.0 and older.
* Fix location bar formatting support in Firefox 50.

Version 2015.11.16.23

Version 2015.10.10.15b
* Fix non-compliant usage of 'let' in Firefox 44 (Bugs 589199, 1202902)
* Rename preference names from status4evar.* to extensions.caligon.s4e.*
* Fix location bar progress padding in Firefox 42.

Version 2015.08.19.02b
* Drop support for Firefox 37.0 and older.
* Fix window.caligon.status4evar is undefined bugs.
* Improve fullscreen HTML5 video detection (fix on Google Drive). Support fullscreen iframe elements.
* Add e10s multi-process support for detecting fullscreen video.

Version 2015.07.14.22b
* Fix location bar formatting support in Firefox 42.

Version 2015.02.06.23
* Fix toolbar pop-up menu bugs by removing legacy Australis compatibility code that is no longer needed.
* Fix location bar origin label offset in Firefox 31 ESR.

Version 2015.01.31.02b
* Drop support for Firefox 30.0 and older. Unify Australis integration.
* Fix Downloads button icon in Firefox 35+.
* Improve Downloads button integration in Menu Panel.
* When the Downloads button is in the Menu panel, or overflow panel, show the download finished animation on the panel button.

Version 2014.07.06.05
* Fix location bar status interactions with domain highlighting.
* Fix origin link shift in Firefox 32 location bar.

Version 2014.06.29.16b
* Fix gripper styling on OS X (Australis)
* Fix Windows download button attention icon (Australis).
* Add throbber toolbar item (Australis).
* Updated translations for: bg-BG, da, de, dsb, el, eo, es-ES, fr, he, hsb, ja-JP, nb-NO, nl, pl, ro, sr, sv-SE, tr, zh-CN, zh-TW

Version 2014.06.01.03b
* Fix download status implementation to be compatible with Firefox 29.
* Fix localization issues. (Australis)
* Improved customization event handling. (Australis)
* Support opening download manager with in-tab UI.
* Support 3rd Party download managers by exposing "status4evar.download.button.action.command" which defines a command name to execute.
* Minimize location bar progress when the location bar is focused.
* Updated translations for: da, de, dsb, el, eo, es-ES, fr, hsb, ja-JP, nl, pl, ro, sr, sv-SE, tr, zh-CN, zh-TW

Version 2014.05.03.06
* Fix download button icon styles. (Australis)
* Fix Windows and Linux status bar styles. OS X Still needs work. (Australis)
* Support legacy status bar panel. (Australis)

Version 2014.04.27.21
* Fix toolbars overlapping download finished animation.
* Australis style fixes.
* Australis first-run widget migration.
* Updated translations for: zh-CN

Version 2014.02.18.00b
* Improve fullscreen HTML5 video detection (fix on YouTube).
* Update license to MPL 2.0.
* Update locales for new strings.
* Updated translations for: sr

Version 2014.01.29.16b
* Add advanced option to hide status popup on fullscreen HTML5 video.
* Fix JS Download API integration.
* Fix status/add-on bar item placement regression from Australis support.
* Updated translations for: ro

Version 2014.01.25.23b
* Initial Australis support.
* Drop support for Firefox 23.0 and older.
* Classic download manager is not available in Firefox 26.0 and newer.

Version 2013.10.31.22
* Updated translations for: es-ES, he

Version 2013.09.14.22b
* Don't use Firefox 22 JavaScript syntax in new Downloads implementation.

Version 2013.09.14.15b
* Support new Downloads.jsm API in Firefox 26+.
* Support UI changes in Firefox 26.
* Update locales for new strings.
* Updated translations for: en-US, pl, sr

Version 2013.05.14.21b
* Fix sticky extension status messages. Instead of simply hiding the status text, actually clear the saved status messages.
* Fix download button panel when the S4E download button is not present on a toolbar.
* Update locales for new strings.
* Updated translations for: el, es-ES, he, ja-JP, tr, zh-TW

Version 2013.05.12.17b
* Add option to control the download button action (Firefox default / panel / library / classic window).
* Add an option to disable the download complete animation.
* Add an option to automatically clear the download complete button highlight.
* Do not highlight the download button if the panel/library/window is already open.

Version 2013.02.16.23
* Fix potential incorrect formatting of the location bar status display after clicking on a link.
* Fix RTL locale support.
* Updated translations for: de, nb-NO, sr, sv-SE, zh-TW

Version 2013.01.16.00b
* Drop support for Firefox 16.0 and older.
* Fix protocol hiding in the toolbar and pop-up.
* Update location bar progress styling.
* Integrate with the new downloads panel.
* Integrate with the new per-window private browsing.
* Updated translations for: bg-BG, cs, da, de, dsb, el, eo, es-ES, fr, he, hsb, it-IT, ja-JP, nb-NO, nl, pt-BR, ro, zh-CN, zh-TW

Version 2012.11.11.16b
* Make compatible with Firefox 19 CSS parser changes.

Version 2012.10.13.20b
* Drop support for Firefox 9.0 and older.

build.sh
#!/bin/bash

XPI_FILE=s4e-revived

path=$(dirname "$(realpath -s "$0")")
temp_path="$path"/src_temp

cd "$path"/src || exit

mkdir "$temp_path"
cp -r "$path"/src/* "$temp_path"/

cd "$temp_path" || exit

cp ../LICENSE ./

mapfile -t locales < <(find "$temp_path"/chrome/locale -maxdepth 1 -exec basename {} \; | sed s/en$// | sed s/LICENSE$// | sed s/LICENSE.in$// | sed s/locale$// | sed -r '/^\s*$/d' | sort -u)

for locale in "${locales[@]}"; do
    extName=$(grep 'name=' "$temp_path"/chrome/locale/"$locale"/meta.properties | cut -d'=' -f2)
    extDesc=$(grep 'description=' "$temp_path"/chrome/locale/"$locale"/meta.properties | cut -d'=' -f2)
    extTrans=$(grep 'translator=' "$temp_path"/chrome/locale/"$locale"/meta.properties | cut -d'=' -f2)
    extContr=$(grep 'em:contributor=' "$temp_path"/install.rdf | cut -d'=' -f2 | sed -e 's/^"//' -e 's/"$//')
    localized+=$(
        cat <<EOF

    <em:localized>
      <Description
        em:locale="$locale"
        em:name="$extName"
        em:description="$extDesc"
        em:translator="$extTrans"
        em:contributor="$extContr"/>
    </em:localized>
\l
EOF
    )
done

localized=$(echo "${localized}" | sed '/^$/d' | sed ':a;N;$!ba;s/\n/\\n/g' | sed 's/\$/\\$/g')
sed -i "s|_localized_|$localized|" "$temp_path"/install.rdf

echo "Creating xpt file"
python2 "$(xdg-user-dir DOWNLOAD)"/firefox-sdk/sdk/bin/typelib.py ./components/status4evar.idl -o ./components/status4evar.xpt -I "$(xdg-user-dir DOWNLOAD)"/firefox-sdk/idl/

echo "Removing old xpi file"
if [ -f "$path/$XPI_FILE.xpi" ]; then
    rm "$path/$XPI_FILE.xpi"
fi

echo "Creating xpi file"
zip -r $XPI_FILE.xpi ./*

cd "$path" || exit
mv "$temp_path/$XPI_FILE".xpi "$path"/
rm -rf "$temp_path"


src/chrome/content/overlay.js
/*
 * ***** BEGIN LICENSE BLOCK *****
 * 
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * 
 * Copyright (C) 2010, 2012-2015 Matthew Turnbull <sparky@bluefang-logic.com>. All Rights Reserved.
 * 
 * ***** END LICENSE BLOCK *****
 *
 * Toolbar popup handling code based on Mozilla Foundation code:
 * https://hg.mozilla.org/mozilla-central/file/6fae9d6feec8/browser/base/content/browser.js#l4114
 * Updated to add gNavToolbox.externalToolbars to toolbarNodes list.
*/

"use strict";

Components.utils.import("resource://status4evar/Australis.jsm");

if(!hawkeye116477) var hawkeye116477 = {};

window.addEventListener("load", function buildS4E()
{
	window.removeEventListener("load", buildS4E, false);

	Components.utils.import("resource://status4evar/Status4Evar.jsm");

	hawkeye116477.status4evar = new Status4Evar(window, gBrowser, gNavToolbox, PanelUI);
	hawkeye116477.status4evar.setup();
}, false);



src/chrome/content/content-thunk.js
/*
 * ***** BEGIN LICENSE BLOCK *****
 * 
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * 
 * Copyright (C) 2015 Matthew Turnbull <sparky@bluefang-logic.com>. All Rights Reserved.
 * 
 * ***** END LICENSE BLOCK *****
*/


function handleVideoDetect(message)
{
	let isVideo = false;

	let fsEl = content.document.mozFullScreenElement;
	if(fsEl)
	{
		isVideo = (
			fsEl.nodeName == "VIDEO"
			|| (fsEl.nodeName == "IFRAME" && fsEl.contentDocument && fsEl.contentDocument.getElementsByTagName("VIDEO").length > 0)
			|| fsEl.getElementsByTagName("VIDEO").length > 0
		);
	}

	sendAsyncMessage("status4evar@hawkeye116477:video-detect-answer", { isVideo: isVideo});
}

addMessageListener("status4evar@hawkeye116477:video-detect", handleVideoDetect);



src/defaults/preferences/defaults.js
/*
 * ***** BEGIN LICENSE BLOCK *****
 * 
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * 
 * Copyright (C) 2010-2015 Matthew Turnbull <sparky@bluefang-logic.com>. All Rights Reserved.
 * 
 * ***** END LICENSE BLOCK *****
*/

pref("extensions.hawkeye116477.s4e.addonbar.borderStyle", false);
pref("extensions.hawkeye116477.s4e.addonbar.legacyShim", true);
pref("extensions.hawkeye116477.s4e.addonbar.windowGripper", true);

pref("extensions.hawkeye116477.s4e.advanced.showWarning", true);
pref("extensions.hawkeye116477.s4e.advanced.status.detectFullScreen", true);
pref("extensions.hawkeye116477.s4e.advanced.status.detectVideo", true);
pref("extensions.hawkeye116477.s4e.advanced.urlbar.forceBinding", false);

pref("extensions.hawkeye116477.s4e.download.button.action", 1);
pref("extensions.hawkeye116477.s4e.download.button.action.command", "");
pref("extensions.hawkeye116477.s4e.download.color.active", "#33FF33");
pref("extensions.hawkeye116477.s4e.download.color.paused", "#808080");
pref("extensions.hawkeye116477.s4e.download.force", true);
pref("extensions.hawkeye116477.s4e.download.label", 0);
pref("extensions.hawkeye116477.s4e.download.label.force", false);
pref("extensions.hawkeye116477.s4e.download.notify.animate", true);
pref("extensions.hawkeye116477.s4e.download.notify.timeout", 60);
pref("extensions.hawkeye116477.s4e.download.progress", 1);
pref("extensions.hawkeye116477.s4e.download.tooltip", 1);

pref("extensions.hawkeye116477.s4e.firstRun", true);
pref("extensions.hawkeye116477.s4e.firstRun.australis", true);

pref("extensions.hawkeye116477.s4e.progress.toolbar.css", "#33FF33");
pref("extensions.hawkeye116477.s4e.progress.toolbar.force", false);
pref("extensions.hawkeye116477.s4e.progress.toolbar.style", false);
pref("extensions.hawkeye116477.s4e.progress.toolbar.style.advanced", false);

pref("extensions.hawkeye116477.s4e.progress.urlbar", 1);
pref("extensions.hawkeye116477.s4e.progress.urlbar.css", "#33FF33");
pref("extensions.hawkeye116477.s4e.progress.urlbar.style", true);
pref("extensions.hawkeye116477.s4e.progress.urlbar.style.advanced", false);

pref("extensions.hawkeye116477.s4e.status", 1);
pref("extensions.hawkeye116477.s4e.status.default", true);
pref("extensions.hawkeye116477.s4e.status.network", true);
pref("extensions.hawkeye116477.s4e.status.network.xhr", true);
pref("extensions.hawkeye116477.s4e.status.timeout", 10);
pref("extensions.hawkeye116477.s4e.status.linkOver", 1);
pref("extensions.hawkeye116477.s4e.status.linkOver.delay.show", 70);
pref("extensions.hawkeye116477.s4e.status.linkOver.delay.hide", 150);

pref("extensions.hawkeye116477.s4e.status.toolbar.maxLength", 0);

pref("extensions.hawkeye116477.s4e.status.urlbar.align", 0);
pref("extensions.hawkeye116477.s4e.status.urlbar.color", "GrayText");
pref("extensions.hawkeye116477.s4e.status.urlbar.position", 33);

pref("extensions.hawkeye116477.s4e.status.popup.invertMirror", false);
pref("extensions.hawkeye116477.s4e.status.popup.mouseMirror", true);



src/components/status4evar.js
/*
 * ***** BEGIN LICENSE BLOCK *****
 * 
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * 
 * Copyright (C) 2010-2015 Matthew Turnbull <sparky@bluefang-logic.com>. All Rights Reserved.
 * 
 * ***** END LICENSE BLOCK *****
*/

"use strict";

// Component constants
const CC = Components.classes;
const CI = Components.interfaces;
const CU = Components.utils;

CU.import("resource://gre/modules/XPCOMUtils.jsm");
CU.import("resource://gre/modules/Services.jsm");

const CURRENT_MIGRATION = 7;

function Status_4_Evar(){}

Status_4_Evar.prototype =
{
	classID:        Components.ID("{13b3595e-7bb5-4cfe-bbfa-82c900a4d7bf}"),
	QueryInterface: XPCOMUtils.generateQI([
	                     CI.nsISupportsWeakReference,
	                     CI.nsIObserver,
	                     CI.nsIStatus4Evar
	                ]),

	prefs:          null,

	addonbarBorderStyle:            false,
	addonbarLegacyShim:             true,
	addonbarWindowGripper:          true,

	advancedStatusDetectFullScreen: true,
	advancedStatusDetectVideo:      true,
	advancedUrlbarForceBinding:     false,

	downloadButtonAction:           1,
	downloadButtonActionCommand:    "",
	downloadColorActive:            null,
	downloadColorPaused:            null,
	downloadForce:                  false,
	downloadLabel:                  0,
	downloadLabelForce:             true,
	downloadNotifyAnimate:          true,
	downloadNotifyTimeout:          60000,
	downloadProgress:               1,
	downloadTooltip:                1,

	firstRun:                       true,
	firstRunAustralis:              true,

	progressToolbarCSS:             null,
	progressToolbarForce:           false,
	progressToolbarStyle:           false,

	progressUrlbar:                 1,
	progressUrlbarCSS:              null,
	progressUrlbarStyle:            true,

	status:                         1,
	statusDefault:                  true,
	statusNetwork:                  true,
	statusTimeout:                  10000,
	statusLinkOver:                 1,
	statusLinkOverDelayShow:        70,
	statusLinkOverDelayHide:        150,

	statusToolbarMaxLength:         0,

	statusUrlbarAlign:              null,
	statusUrlbarColor:              null,
	statusUrlbarPosition:           33,

	statusUrlbarInvertMirror:       false,
	statusUrlbarMouseMirror:        true,

	pref_registry:
	{
		"addonbar.borderStyle":
		{
			update: function()
			{
				this.addonbarBorderStyle = this.prefs.getBoolPref("addonbar.borderStyle");
			},
			updateWindow: function(win)
			{
				let browser_bottom_box = win.hawkeye116477.status4evar.getters.browserBottomBox;
				if(browser_bottom_box)
				{
					this.setBoolElementAttribute(browser_bottom_box, "s4eboarder", this.addonbarBorderStyle);
				}
			}
		},

		"addonbar.legacyShim":
		{
			update: function()
			{
				this.addonbarLegacyShim = this.prefs.getBoolPref("addonbar.legacyShim");

				CU.import("resource://status4evar/Australis.jsm", {}).AustralisTools.updateLegacyShim(this.addonbarLegacyShim);
			}
		},

		"addonbar.windowGripper":
		{
			update: function()
			{
				this.addonbarWindowGripper = this.prefs.getBoolPref("addonbar.windowGripper");
			},
			updateWindow: function(win)
			{
				win.hawkeye116477.status4evar.toolbars.updateWindowGripper(true);
			}
		},

		"advanced.status.detectFullScreen":
		{
			update: function()
			{
				this.advancedStatusDetectFullScreen = this.prefs.getBoolPref("advanced.status.detectFullScreen");
			}
		},

		"advanced.status.detectVideo":
		{
			update: function()
			{
				this.advancedStatusDetectVideo = this.prefs.getBoolPref("advanced.status.detectVideo");
			}
		},

		"advanced.urlbar.forceBinding":
		{
			update: function()
			{
				this.advancedUrlbarForceBinding = this.prefs.getBoolPref("advanced.urlbar.forceBinding");
			},
			updateWindow: function(win)
			{
				let urlbar = win.hawkeye116477.status4evar.getters.urlbar;
				if(urlbar)
				{
					this.setBoolElementAttribute(urlbar, "s4eforce", this.advancedUrlbarForceBinding);
				}
			}
		},

		"download.button.action":
		{
			update: function()
			{
				this.downloadButtonAction = this.prefs.getIntPref("download.button.action");
			},
			updateWindow: function(win)
			{
				win.hawkeye116477.status4evar.downloadStatus.updateBinding();
			}
		},

		"download.button.action.command":
		{
			update: function()
			{
				this.downloadButtonActionCommand = this.prefs.getCharPref("download.button.action.command");
			}
		},

		"download.color.active":
		{
			update: function()
			{
				this.downloadColorActive = this.prefs.getCharPref("download.color.active");
			},
			updateDynamicStyle: function(sheet)
			{
				sheet.cssRules[4].style.backgroundColor = this.downloadColorActive;
			}
		},

		"download.color.paused":
		{
			update: function()
			{
				this.downloadColorPaused = this.prefs.getCharPref("download.color.paused");
			},
			updateDynamicStyle: function(sheet)
			{
				sheet.cssRules[5].style.backgroundColor = this.downloadColorPaused;
			}
		},

		"download.force":
		{
			update: function()
			{
				this.downloadForce = this.prefs.getBoolPref("download.force");
			},
			updateWindow: function(win)
			{
				let download_button = win.hawkeye116477.status4evar.getters.downloadButton;
				if(download_button)
				{
					this.setBoolElementAttribute(download_button, "forcevisible", this.downloadForce);
				}
			}
		},

		"download.label":
		{
			update: function()
			{
				this.downloadLabel = this.prefs.getIntPref("download.label");
			},
			updateWindow: function(win)
			{
				win.hawkeye116477.status4evar.downloadStatus.updateButton();
			}
		},

		"download.label.force":
		{
			update: function()
			{
				this.downloadLabelForce = this.prefs.getBoolPref("download.label.force");
			},
			updateWindow: function(win)
			{
				let download_button = win.hawkeye116477.status4evar.getters.downloadButton;
				if(download_button)
				{
					this.setBoolElementAttribute(download_button, "forcelabel", this.downloadLabelForce);
				}
			}
		},

		"download.notify.animate":
		{
			update: function()
			{
				this.downloadNotifyAnimate = this.prefs.getBoolPref("download.notify.animate");
			}
		},

		"download.notify.timeout":
		{
			update: function()
			{
				this.downloadNotifyTimeout = (this.prefs.getIntPref("download.notify.timeout") * 1000);
			}
		},

		"download.progress":
		{
			update: function()
			{
				this.downloadProgress = this.prefs.getIntPref("download.progress");
			},
			updateWindow: function(win)
			{
				win.hawkeye116477.status4evar.downloadStatus.updateButton();
			}
		},

		"download.tooltip":
		{
			update: function()
			{
				this.downloadTooltip = this.prefs.getIntPref("download.tooltip");
			},
			updateWindow: function(win)
			{
				win.hawkeye116477.status4evar.downloadStatus.updateButton();
			}
		},

		"progress.toolbar.css":
		{
			update: function()
			{
				this.progressToolbarCSS = this.prefs.getCharPref("progress.toolbar.css");
			},
			updateDynamicStyle: function(sheet)
			{
				sheet.cssRules[2].style.background = this.progressToolbarCSS;
			}
		},

		"progress.toolbar.force":
		{
			update: function()
			{
				this.progressToolbarForce = this.prefs.getBoolPref("progress.toolbar.force");
			},
			updateWindow: function(win)
			{
				let toolbar_progress = win.hawkeye116477.status4evar.getters.toolbarProgress;
				if(toolbar_progress)
				{
					this.setBoolElementAttribute(toolbar_progress, "forcevisible", this.progressToolbarForce);
				}
			}
		},

		"progress.toolbar.style":
		{
			update: function()
			{
				this.progressToolbarStyle = this.prefs.getBoolPref("progress.toolbar.style");
			},
			updateWindow: function(win)
			{
				let toolbar_progress = win.hawkeye116477.status4evar.getters.toolbarProgress;
				if(toolbar_progress)
				{
					this.setBoolElementAttribute(toolbar_progress, "s4estyle", this.progressToolbarStyle);
				}
			}
		},

		"progress.urlbar":
		{
			update: function()
			{
				switch(this.prefs.getIntPref("progress.urlbar"))
				{
					case 0:
						this.progressUrlbar = null;
						break;
					case 1:
						this.progressUrlbar = "end";
						break;
					case 2:
						this.progressUrlbar = "begin";
						break;
					default:
						this.progressUrlbar = "center";
						break;
				}
			},
			updateWindow: function(win)
			{
				let urlbar = win.hawkeye116477.status4evar.getters.urlbar;
				let urlbar_progress = win.hawkeye116477.status4evar.getters.urlbarProgress;
				if(urlbar && urlbar_progress)
				{
					if(this.progressUrlbar)
					{
						urlbar.pmpack = this.progressUrlbar;
					}
					urlbar_progress.hidden = !this.progressUrlbar;
				}
			}
		},

		"progress.urlbar.css":
		{
			update: function()
			{
				this.progressUrlbarCSS = this.prefs.getCharPref("progress.urlbar.css");
			},
			updateDynamicStyle: function(sheet)
			{
				sheet.cssRules[1].style.background = this.progressUrlbarCSS;
			}
		},

		"progress.urlbar.style":
		{
			update: function()
			{
				this.progressUrlbarStyle = this.prefs.getBoolPref("progress.urlbar.style");
			},
			updateWindow: function(win)
			{
				let urlbar = win.hawkeye116477.status4evar.getters.urlbar;
				if(urlbar)
				{
					this.setBoolElementAttribute(urlbar, "s4estyle", this.progressUrlbarStyle);
				}
			}
		},

		"status":
		{
			update: function()
			{
				this.status = this.prefs.getIntPref("status");
			},
			updateWindow: function(win)
			{
				win.hawkeye116477.status4evar.statusService.clearStatusField();
				win.hawkeye116477.status4evar.statusService.updateStatusField(true);
			}
		},

		"status.default":
		{
			update: function()
			{
				this.statusDefault = this.prefs.getBoolPref("status.default");
			},
			updateWindow: function(win)
			{
				win.hawkeye116477.status4evar.statusService.buildTextOrder();
				win.hawkeye116477.status4evar.statusService.updateStatusField(true);
			}
		},

		"status.linkOver":
		{
			update: function()
			{
				this.statusLinkOver = this.prefs.getIntPref("status.linkOver");
			}
		},

		"status.linkOver.delay.show":
		{
			update: function()
			{
				this.statusLinkOverDelayShow = this.prefs.getIntPref("status.linkOver.delay.show");
			}
		},

		"status.linkOver.delay.hide":
		{
			update: function()
			{
				this.statusLinkOverDelayHide = this.prefs.getIntPref("status.linkOver.delay.hide");
			}
		},

		"status.network":
		{
			update: function()
			{
				this.statusNetwork = this.prefs.getBoolPref("status.network");
			},
			updateWindow: function(win)
			{
				win.hawkeye116477.status4evar.statusService.buildTextOrder();
			}
		},

		"status.network.xhr":
		{
			update: function()
			{
				this.statusNetworkXHR = this.prefs.getBoolPref("status.network.xhr");
			},
			updateWindow: function(win)
			{
				win.hawkeye116477.status4evar.statusService.buildTextOrder();
			}
		},

		"status.popup.invertMirror":
		{
			update: function()
			{
				this.statusUrlbarInvertMirror = this.prefs.getBoolPref("status.popup.invertMirror");
			},
			updateWindow: function(win)
			{
				let statusOverlay = win.hawkeye116477.status4evar.getters.statusOverlay;
				if(statusOverlay)
				{
					statusOverlay.invertMirror = this.statusUrlbarInvertMirror;
				}
			}
		},

		"status.popup.mouseMirror":
		{
			update: function()
			{
				this.statusUrlbarMouseMirror = this.prefs.getBoolPref("status.popup.mouseMirror");
			},
			updateWindow: function(win)
			{
				let statusOverlay = win.hawkeye116477.status4evar.getters.statusOverlay;
				if(statusOverlay)
				{
					statusOverlay.mouseMirror = this.statusUrlbarMouseMirror;
				}
			}
		},

		"status.timeout":
		{
			update: function()
			{
				this.statusTimeout = (this.prefs.getIntPref("status.timeout") * 1000);
			},
			updateWindow: function(win)
			{
				win.hawkeye116477.status4evar.statusService.updateStatusField(true);
			}
		},

		"status.toolbar.maxLength":
		{
			update: function()
			{
				this.statusToolbarMaxLength = this.prefs.getIntPref("status.toolbar.maxLength");
			},
			updateWindow: function(win)
			{
				let status_widget = win.hawkeye116477.status4evar.getters.statusWidget;
				if(status_widget)
				{
					status_widget.maxWidth = (this.statusToolbarMaxLength || "");
				}
			}
		},

		"status.urlbar.align":
		{
			update: function()
			{
				switch(this.prefs.getIntPref("status.urlbar.align"))
				{
					case 0:
						this.statusUrlbarAlign = null;
						break;
					case 1:
						this.statusUrlbarAlign = "left";
						break;
					default:
						this.statusUrlbarAlign = "absolute";
						break;
				}
			},
			updateWindow: function(win)
			{
				let urlbar = win.hawkeye116477.status4evar.getters.urlbar;
				if(urlbar)
				{
					urlbar.s4esalign = this.statusUrlbarAlign;
					urlbar.updateOverLinkLayout();
				}
			}
		},

		"status.urlbar.color":
		{
			update: function()
			{
				this.statusUrlbarColor = this.prefs.getCharPref("status.urlbar.color");
			},
			updateDynamicStyle: function(sheet)
			{
				sheet.cssRules[3].style.color = this.statusUrlbarColor;
			}
		},

		"status.urlbar.position":
		{
			update: function()
			{
				this.statusUrlbarPosition = this.prefs.getIntPref("status.urlbar.position");

				if(this.statusUrlbarPosition < 10)
				{
					this.statusUrlbarPosition = 10;
				}
				else if(this.statusUrlbarPosition > 90)
				{
					this.statusUrlbarPosition = 90;
				}
			},
			updateWindow: function(win)
			{
				let urlbar = win.hawkeye116477.status4evar.getters.urlbar;
				if(urlbar)
				{
					urlbar.s4espos = this.statusUrlbarPosition;
					urlbar.updateOverLinkLayout();
				}
			}
		}
	},

	// nsIObserver
	observe: function(subject, topic, data)
	{
		try
		{
			switch(topic)
			{
				case "profile-after-change":
					this.startup();
					break;
				case "quit-application":
					this.shutdown();
					break;
				case "nsPref:changed":
					this.updatePref(data, true);
					break;
			}
		}
		catch(e)
		{
			CU.reportError(e);
		}
	},

	startup: function()
	{
		this.prefs = Services.prefs.getBranch("extensions.hawkeye116477.s4e.").QueryInterface(CI.nsIPrefBranch2);

		this.migratePrefsRoot();

		this.firstRun = this.prefs.getBoolPref("firstRun");
		if(this.firstRun)
		{
			this.prefs.setBoolPref("firstRun", false);
		}

		this.firstRunAustralis = this.prefs.getBoolPref("firstRun.australis");
		if(this.firstRunAustralis && Services.vc.compare("28.*", Services.appinfo.version) < 0)
		{
			this.prefs.setBoolPref("firstRun.australis", false);
		}

		this.migrate();

		for(let pref in this.pref_registry)
		{
			let pro = this.pref_registry[pref];

			pro.update = pro.update.bind(this);
			if(pro.updateWindow)
			{
				pro.updateWindow = pro.updateWindow.bind(this);
			}
			if(pro.updateDynamicStyle)
			{
				pro.updateDynamicStyle = pro.updateDynamicStyle.bind(this);
			}

			this.prefs.addObserver(pref, this, true);

			this.updatePref(pref, false);
		}

		Services.obs.addObserver(this, "quit-application", true);
	},

	shutdown: function()
	{
		Services.obs.removeObserver(this, "quit-application");

		for(let pref in this.pref_registry)
		{
			this.prefs.removeObserver(pref, this);
		}

		this.prefs = null;
	},

	migratePrefsRoot: function()
	{
		let oldPrefs = Services.prefs.getBranch("status4evar.").QueryInterface(CI.nsIPrefBranch2);

		let childPrefs = oldPrefs.getChildList("");
		childPrefs.forEach(function(pref)
		{
			if(oldPrefs.prefHasUserValue(pref))
			{
				switch(oldPrefs.getPrefType(pref))
				{
					case CI.nsIPrefBranch2.PREF_STRING:
						this.prefs.setCharPref(pref, oldPrefs.getCharPref(pref));
						break;
					case CI.nsIPrefBranch2.PREF_INT:
						this.prefs.setIntPref(pref, oldPrefs.getIntPref(pref));
						break;
					case CI.nsIPrefBranch2.PREF_BOOL:
						this.prefs.setBoolPref(pref, oldPrefs.getBoolPref(pref));
						break;
				}

				oldPrefs.clearUserPref(pref);
			}
		}, this);
	},

	migrate: function()
	{
		if(!this.firstRun)
		{
			let migration = 0;
			try
			{
				migration = this.prefs.getIntPref("migration");
			}
			catch(e) {}

			switch(migration)
			{
				case 5:
					this.migrateBoolPref("status.detectFullScreen", "advanced.status.detectFullScreen");
					break;
				case 6:
					let oldDownloadAction = this.prefs.getIntPref("download.button.action");
					let newDownloadAction = 1;
					switch(oldDownloadAction)
					{
						case 2:
							newDownloadAction = 1;
							break;
						case 3:
							newDownloadAction = 2;
							break;
						case 4:
							newDownloadAction = 1;
							break;
					}
					this.prefs.setIntPref("download.button.action", newDownloadAction);

					if(oldDownloadAction == 4 && Services.vc.compare("26.0", Services.appinfo.version) > 0)
					{
						this.prefs.setBoolPref("browser.download.useToolkitUI", true);
					}
					break;
				case CURRENT_MIGRATION:
					break;
			}
		}

		this.prefs.setIntPref("migration", CURRENT_MIGRATION);
	},

	migrateBoolPref: function(oldPref, newPref)
	{
		if(this.prefs.prefHasUserValue(oldPref))
		{
			this.prefs.setBoolPref(newPref, this.prefs.getBoolPref(oldPref));
			this.prefs.clearUserPref(oldPref);
		}
	},

	migrateIntPref: function(oldPref, newPref)
	{
		if(this.prefs.prefHasUserValue(oldPref))
		{
			this.prefs.setIntPref(newPref, this.prefs.getIntPref(oldPref));
			this.prefs.clearUserPref(oldPref);
		}
	},

	migrateCharPref: function(oldPref, newPref)
	{
		if(this.prefs.prefHasUserValue(oldPref))
		{
			this.prefs.setCharPref(newPref, this.prefs.getCharPref(oldPref));
			this.prefs.clearUserPref(oldPref);
		}
	},

	updatePref: function(pref, updateWindows)
	{
		if(!(pref in this.pref_registry))
		{
			return;
		}
		let pro = this.pref_registry[pref];

		pro.update();

		if(updateWindows)
		{
			let windowsEnum = Services.wm.getEnumerator("navigator:browser");
			while(windowsEnum.hasMoreElements())
			{
				this.updateWindow(windowsEnum.getNext(), pro);
			}
		}

		if(pro.alsoUpdate)
		{
			pro.alsoUpdate.forEach(function (alsoPref)
			{
				this.updatePref(alsoPref);
			}, this);
		}
	},

	// Updtate a browser window
	updateWindow: function(win, pro)
	{
		if(!(win instanceof CI.nsIDOMWindow)
		|| !(win.document.documentElement.getAttribute("windowtype") == "navigator:browser"))
		{
			return;
		}

		if(pro)
		{
			this.handlePro(win, pro);
		}
		else
		{
			for(let pref in this.pref_registry)
			{
				this.handlePro(win, this.pref_registry[pref]);
			}
		}
	},

	handlePro: function(win, pro)
	{
		if(pro.updateWindow)
		{
			pro.updateWindow(win);
		}

		if(pro.updateDynamicStyle)
		{
			let styleSheets = win.document.styleSheets;
			for(let i = 0; i < styleSheets.length; i++)
			{
				let styleSheet = styleSheets[i];
				if(styleSheet.href == "chrome://status4evar/skin/dynamic.css")
				{
					pro.updateDynamicStyle(styleSheet);
					break;
				}
			}
		}
	},

	setBoolElementAttribute: function(elem, attr, val)
	{
		if(val)
		{
			elem.setAttribute(attr, "true");
		}
		else
		{
			elem.removeAttribute(attr);
		}
	},

	setStringElementAttribute: function(elem, attr, val)
	{
		if(val)
		{
			elem.setAttribute(attr, val);
		}
		else
		{
			elem.removeAttribute(attr);
		}
	},

	resetPrefs: function()
	{
		let childPrefs = this.prefs.getChildList("");
		childPrefs.forEach(function(pref)
		{
			if(this.prefs.prefHasUserValue(pref))
			{
				this.prefs.clearUserPref(pref);
			}
		}, this);
	}
};

const NSGetFactory = XPCOMUtils.generateNSGetFactory([Status_4_Evar]);

